pg-extend/src/pg_datum.rs

```diff
@@ -370,6 +363,101 @@ where
     }
 }
 
+/// Inner trait used to limit which types can be used for direct casting
+#[doc(hidden)]
+pub trait PgPrimitiveDatum {}
+
+impl PgPrimitiveDatum for i16 {}
+impl PgPrimitiveDatum for i32 {}
+impl PgPrimitiveDatum for i64 {}
+impl PgPrimitiveDatum for f32 {}
+impl PgPrimitiveDatum for f64 {}
+
+impl<'s, T> TryFromPgDatum<'s> for &[T]
+where
+    T: 's + TryFromPgDatum<'s> + PgPrimitiveDatum,
+{
+    fn try_from<'mc>(_: &'mc PgAllocator, datum: PgDatum<'mc>) -> Result<Self, &'static str>
+    where
+        Self: 's,
+        'mc: 's,
+    {
+        if let Some(datum) = datum.0 {
+            unsafe {
+                let datum = datum as *mut pg_sys::varlena;
+                if datum.is_null() {
+                    return Err("datum was NULL");
+                }
+
+                #[allow(clippy::cast_ptr_alignment)]
+                let arr_type = pg_sys::pg_detoast_datum(datum) as *mut pg_sys::ArrayType;
```

Comment:
This might allocate via `palloc()` depending on if the incoming Datum was TOASTed or not.

You can compare pointers `datum == arr_type` to see if it allocated or not.  If it did, then you'd want to `pfree()` `arr_type` whenever this thing gets dropped